pkg/kubelet/pod_manager.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/metrics"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/golang/glog"
+)
+
+type podManager struct {
+	// Protects all internal pod storage/mappings.
+	lock sync.RWMutex
+	pods []api.Pod
+	// Record the set of mirror pods (see mirror_manager.go for more details);
+	// similar to pods, this is not immutable and is protected by the same podLock.
+	// Note that Kubelet.pods do not contain mirror pods as they are filtered
+	// out beforehand.
+	mirrorPods mirrorPods
+
+	// A mirror pod manager which provides helper functions.
+	mirrorManager mirrorManager
+}
+
+func newPodManager(apiserverClient client.Interface) *podManager {
+	podManager := podManager{}
+	podManager.mirrorManager = newBasicMirrorManager(apiserverClient)
+	podManager.mirrorPods = *newMirrorPods()
+	podManager.pods = []api.Pod{}
+	return &podManager
+}
+
+func (self *podManager) setPods(pods []api.Pod) {
+	self.pods, self.mirrorPods = filterAndCategorizePods(pods)
+}
+
+// Update the internal pods with those provided by the update.
+// Records new and updated pods in newPods and updatedPods.
```

Comment:
Sorry, this one is unfixed from one of my PRs: instead of newPods and updatedPods should be podSyncTypes